lib/ddtrace/context_provider.rb

```diff
@@ -28,18 +28,23 @@ class ThreadLocalContext
     # a different \Context for each thread. In synchronous tracer, this
     # is required to prevent multiple threads sharing the same \Context
     # in different executions.
+    #
+    # To support multiple tracers simultaneously, each \ThreadLocalContext
+    # instance has its own thread-local variable.
     def initialize
+      @key = "datadog_context_#{object_id}".to_sym
+
       self.local = Datadog::Context.new
     end
 
     # Override the thread-local context with a new context.
     def local=(ctx)
-      Thread.current[:datadog_context] = ctx
+      Thread.current[@key] = ctx
```

Comment:
Is it possible that on the one thread, multiple contexts might be used over time?

I'm trying to see if there's some possible path for contexts to hang around in the thread local memory in a way that would cause a leak. Given we have a number of integrations that reset the context, and others that overwrite it for the purposes of distributed tracing, I'm wondering if either of those could cause a problem like this.

Reply:
@delner `ThreadLocalContext` instances are currently strongly tied to `Tracer` instance: they are created at `Tracer#initialize` and all request for "current thread context" go though, either directly or indirectly, a tracer instance.

For applications configuring `ddtrace` today nothing will change: only one context will exist per thread, and only if a span was created during that thread's execution.

For applications that call `Datadog.configure` multiple times, nothing changes, as `Tracer#initialize` is still only called once, even though `Tracer#configure` is called many times. Currently, `Tracer#configure` does not reconfigure the `context_provider`.

For applications that manually invoke `Tracer.new`, those will have multiple contexts per thread: one `Context` per thread, per tracer instance maximum.

I considered creating a `#shutdown!` method to `ThreadLocalContext` (trigger from `Tracer#shutdown!`), but that would only clean up the current thread's `Context`. I'm not sure if there's a feasible way to clean up another thread's thread-local variable.

I guess this is one of those situations where the drawbacks of thread-local variables come back to bite us 🤔.
Managing the context as an explicit resource is the "ideal" option, albeit a much larger endeavour at this point:
```ruby
tracer1.call_context do |ctx1| # Initialize thread-local context for tracer 1
  tracer2.call_context do |ctx2| # Initialize thread-local context for tracer 2
    ...
  end # Remove thread-local variable for tracer 2's context

  tracer1.call_context do |ctx1| # Thread-local context for tracer 1 already exists, nothing to do
    ...
  end # I did not create the thread-local context, nothing to do
  ...
end # Remove thread-local variable for tracer 1's context
```

This would requiring wrapping all usages of the current context, so quite a lot of work, but it would make managing the lifetime cycle much clearer.

Reply:
I would say the use case for creating multiple tracers + trace local contexts is pretty minimal. Maybe in a testing env there might be a few hundred created, but I'd say in an application there is probably max of 2 per-thread (default one + one manually created by the customer?).

Also, as far as shutdown goes, if the thread closes, the thread local data should be cleaned up as well, so we might not need to try and be fancy here.

All of this hinges on the idea that a minimal number of tracers + thread local contexts exist per-thread.